core/cap-0021.md

```diff
@@ -286,27 +261,41 @@ T2 have the same sequence number but T1 has twice the
 `fee`--validators seeing both should preferentially forward T1, but
 also accept T2 in nominated blocks.
 
-All transactions are validated, sequence numbers increased, and fees
-deducted before any operations are executed.  One consequence is that
-`BUMP_SEQUENCE` operations, though they update an account's `seqTime`
-and `seqLedger`, do not affect the validity of other transactions in
-the same block.
-
-### Transaction results
-
-Because `GeneralPreconditions` specifies multiple pre-conditions,
-there may be multiple reasons why a transaction is invalid.  If any of
-the reasons is permanent---for instance the `maxTime` or `maxLedger`
-has been exceeded--then the transaction is rejected with
+### Transaction validation
+
+The `Preconditions` in each transaction are validated twice.
+
+The first validation occurs when checking that a block itself is valid
+and can be nominated by the consensus protocol.  As part of the
+validation, for each `sourceAccount` there can be only one transaction
+with a given sequence number and the `seqNum` fields must either be
+consecutive or any gaps must be permitted by non-NULL `minSeqNum`
+fields.  All but the transaction with the lowest `seqNum` on a given
+`sourceAccount` must have 0 for the `minSeqAge` and `minSeqLedgerGap`
+fields.
+
+Once a block is externalized by the consensus algorithm, the block is
+applied.  Before executing any operations, fees are charged to the
+source account for all transactions.  Once fees have been deducted
+from all accounts, transactions are one-by-one validated a second time
+then executed.  It is possible for a previously valid transaction to
+fail the second validation, for instance if a `BUMP_SEQUENCE`
+operation made the sequence number invalid.  Whenever a transaction
+fails validation during execution, the `sourceAccount` loses the fee.
+
+Because `TxConditionsV2` specifies multiple pre-conditions, there may
+be multiple reasons why a transaction is invalid.  If any of the
+reasons is permanent---for instance the `maxTime` or `maxLedger` has
+been exceeded--then the transaction is rejected with
 `TransactionResultCode` `txTOO_LATE`.  Otherwise, if all the
 unsatisfied preconditions are based on values that could later be
 valid, then the transaction is rejected with `txTOO_EARLY`.
```

Comment:
It is worth noting that the only precondition that causes too late is the ledger bounds or time bounds preconditions. The vast majority will cause too early. The extra signers will cause bad auth. We should probably specify here that extra signers will cause bad auth not too early/late.